File: smalivm/src/main/java/org/cf/smalivm/opcode/NewInstanceOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.VirtualMachine;
import org.cf.smalivm.type.VirtualClass;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.formats.Instruction21c;
import org.jf.dexlib2.iface.reference.TypeReference;

public class NewInstanceOpFactory implements OpFactory {

    @Override
    public Op create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation, VirtualMachine vm) {
        MethodLocation child = Utils.getNextLocation(location, addressToLocation);
        Instruction21c instr = (Instruction21c) location.getInstruction();
        int destRegister = instr.getRegisterA();
        TypeReference typeRef = (TypeReference) instr.getReference();
        String className = typeRef.getType();
        VirtualClass virtualClass = vm.getClassManager().getVirtualClass(className);

        return new NewInstanceOp(location, child, destRegister, virtualClass, vm);
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/NewInstanceOpFactory.kt
package org.cf.smalivm.opcode

import gnu.trove.map.TIntObjectMap
import org.cf.smalivm.VirtualMachine
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction21c
import org.jf.dexlib2.iface.reference.TypeReference

class NewInstanceOpFactory : OpFactory {
    fun create(location: MethodLocation, addressToLocation: TIntObjectMap<MethodLocation?>?, vm: VirtualMachine): Op {
        val child = Utils.getNextLocation(location, addressToLocation)
        val instr = location.instruction as Instruction21c?
        val destRegister = instr!!.registerA
        val typeRef = instr.reference as TypeReference
        val className = typeRef.type
        val virtualClass = vm.classManager.getVirtualClass(className)
        return NewInstanceOp(location, child, destRegister, virtualClass, vm)
    }
}
